src/octree/mod.rs

```diff
@@ -238,13 +262,98 @@ impl OnDiskOctree {
                 },
             );
         }
-        Ok(OnDiskOctree { meta, nodes })
+
+        Ok(Octree {
+            meta,
+            nodes,
+            data_provider,
+        })
     }
 
-    pub fn new<P: AsRef<Path>>(directory: P) -> Result<Self> {
-        let directory = directory.as_ref().to_owned();
-        let meta_proto = read_meta_proto(&directory)?;
-        Self::from_meta(meta_proto, directory)
+    pub fn get_visible_nodes(&self, projection_matrix: &Matrix4<f32>) -> Vec<NodeId> {
+        let frustum = Frustum::from_matrix4(*projection_matrix).unwrap();
+        let mut open = BinaryHeap::new();
+        maybe_push_node(
+            &mut open,
+            &self.nodes,
+            Relation::Cross,
+            Node::root_with_bounding_cube(Cube::bounding(&self.meta.bounding_box)),
+            projection_matrix,
+        );
+
+        let mut visible = Vec::new();
+        while let Some(current) = open.pop() {
+            match current.relation {
+                Relation::Cross => {
+                    for child_index in 0..8 {
+                        let child = current.node.get_child(ChildIndex::from_u8(child_index));
+                        let child_relation = frustum.contains(&child.bounding_cube.to_aabb3());
+                        if child_relation == Relation::Out {
+                            continue;
+                        }
+                        maybe_push_node(
+                            &mut open,
+                            &self.nodes,
+                            child_relation,
+                            child,
+                            projection_matrix,
+                        );
+                    }
+                }
+                Relation::In => {
+                    // When the parent is fully in the frustum, so are the children.
+                    for child_index in 0..8 {
+                        maybe_push_node(
+                            &mut open,
+                            &self.nodes,
+                            Relation::In,
+                            current.node.get_child(ChildIndex::from_u8(child_index)),
+                            projection_matrix,
+                        );
+                    }
+                }
+                Relation::Out => {
+                    // This should never happen.
+                    unreachable!();
+                }
+            };
+            visible.push(current.node.id);
+        }
+        visible
+    }
+
+    pub fn get_node_data(&self, node_id: &NodeId) -> Result<NodeData> {
+        // TODO(hrapp): If we'd randomize the points while writing, we could just read the
+        // first N points instead of reading everything and skipping over a few.
+        let mut position_color_reads = self
+            .data_provider
+            .data(node_id, vec![node::POSITION_EXT, node::COLOR_EXT])?;
+
+        let get_data = |all_reads: &mut HashMap<String, Box<dyn Read>>,
```

Comment:
you actually don't have to pass in `all_reads` because the closure can access `position_color_reads` anyway. If you still want to pass it in because of readability, you can convert the closure into a normal function `fn get_data(...)` which can still be defined inside of `fn get_node_data()`. If a closure doesn't capture any variable from its environment, it's more of a matter of taste whether to use a closure or a function.

Reply:
Done, removed it from the parameters list.